File: myLeetCode/Array/977. Squares of a Sorted Array.cpp
// 977. Squares of a Sorted Array
// https://leetcode.com/problems/squares-of-a-sorted-array/

#include <algorithm>
#include <iostream>
#include <vector>

class Solution {
  public:
    std::vector<int> sortedSquares_Double_pointer(std::vector<int> &nums) {
        int n = nums.size();
        std::vector<int> ans(n);
        for (int left = 0, right = n - 1, pos = n - 1; left <= right;) {
            if (nums[left] * nums[left] > nums[right] * nums[right]) {
                ans[pos] = nums[left] * nums[left];
                ++left;
            } else {
                ans[pos] = nums[right] * nums[right];
                --right;
            }
            --pos;
        }
        return ans;
    }
};

int main() {
    Solution sol;
    std::vector<int> nums = {-4, -1, 0, 3, 10};
    std::vector<int> ans = sol.sortedSquares_Double_pointer(nums);
    for (int num : ans) {
        std::cout << num << " ";
    }
    return 0;
}

File: myLeetCode/Array/977. Squares of a Sorted Array.py
# 977. Squares of a Sorted Array
# https://leetcode.com/problems/squares-of-a-sorted-array/

class Solution:
    def sortedSquares_Double_pointer(self, nums: list[int]) -> list[int]:
        n = len(nums)
        ans = [0] * n
        left, right = 0, n - 1
        pos = n - 1
        while left <= right:
            if nums[left] * nums[left] > nums[right] * nums[right]:
                ans[pos] = nums[left] * nums[left]
                left += 1
            else:
                ans[pos] = nums[right] * nums[right]
                right -= 1
            pos -= 1
        return ans

def main():
    sol = Solution()
    nums = [-4, -1, 0, 3, 10]
    ans = sol.sortedSquares_Double_pointer(nums)
    print(*ans)
    
if __name__ == "__main__":
    main()


